File: LinkedList/DoublyLinkedList/DoublyLinkedList.c
#include <stdio.h>
#include <stdlib.h>

typedef struct Node {
    int value;
    struct Node* next;
    struct Node* prev;
} Node;

Node* createNode(int value) {
    Node* newNode = (Node*)malloc(sizeof(Node));
    newNode->value = value;
    newNode->next = NULL;
    newNode->prev = NULL;
    return newNode;
}

void prepend(Node** head, int value) {
    Node* newNode = createNode(value);
    if (*head != NULL) {
        (*head)->prev = newNode;
        newNode->next = *head;
    }
    *head = newNode;
}

void append(Node** head, int value) {
    Node* newNode = createNode(value);
    if (*head == NULL) {
        *head = newNode;
        return;
    }
    Node* current = *head;
    while (current->next != NULL) {
        current = current->next;
    }
    current->next = newNode;
    newNode->prev = current;
}

void insert(Node** head, int value, int position) {
    if (position == 0) {
        prepend(head, value);
        return;
    }
    Node* newNode = createNode(value);
    Node* current = *head;
    for (int i = 0; i < position - 1; i++) {
        if (current == NULL) return;
        current = current->next;
    }
    if (current == NULL) return;

    newNode->next = current->next;
    newNode->prev = current;
    if (current->next != NULL) {
        current->next->prev = newNode;
    }
    current->next = newNode;
}

void deleteFirst(Node** head) {
    if (*head == NULL) return;
    Node* temp = *head;
    *head = (*head)->next;
    if (*head != NULL) {
        (*head)->prev = NULL;
    }
    free(temp);
}

void deleteLast(Node** head) {
    if (*head == NULL) return;
    if ((*head)->next == NULL) {
        free(*head);
        *head = NULL;
        return;
    }
    Node* current = *head;
    while (current->next != NULL) {
        current = current->next;
    }
    current->prev->next = NULL;
    free(current);
}

void delete(Node** head, int position) {
    if (*head == NULL) return;
    if (position == 0) {
        deleteFirst(head);
        return;
    }
    Node* current = *head;
    for (int i = 0; i < position; i++) {
        if (current == NULL) return;
        current = current->next;
    }
    if (current == NULL) return;

    if (current->prev != NULL) {
        current->prev->next = current->next;
    }
    if (current->next != NULL) {
        current->next->prev = current->prev;
    }
    free(current);
}

int search(Node* head, int value) {
    Node* current = head;
    int position = 0;
    while (current != NULL) {
        if (current->value == value) return position;
        current = current->next;
        position++;
    }
    return -1;
}

void print(Node* head) {
    Node* current = head;
    if (current == NULL) {
        printf("Empty List\n");
        return;
    }
    while (current != NULL) {
        printf("%d", current->value);
        if (current->next != NULL) printf(" <-> ");
        current = current->next;
    }
    printf("\n");
}

int main() {
    Node* head = NULL;
    append(&head, 10);
    append(&head, 20);
    prepend(&head, 5);
    insert(&head, 15, 2);
    print(head); // 5 <-> 10 <-> 15 <-> 20
    delete(&head, 2);
    print(head); // 5 <-> 10 <-> 20
    return 0;
}

File: LinkedList/DoublyLinkedList/DoublyLinkedList.py
class Node:
    def __init__(self, value):
        self.value = value
        self.next = None
        self.prev = None

class DoublyLinkedList:
    def __init__(self):
        self.head = None

    def prepend(self, value):
        new_node = Node(value)
        if self.head:
            self.head.prev = new_node
            new_node.next = self.head
        self.head = new_node

    def append(self, value):
        new_node = Node(value)
        if not self.head:
            self.head = new_node
            return
        current = self.head
        while current.next:
            current = current.next
        current.next = new_node
        new_node.prev = current

    def insert(self, value, position):
        if position == 0:
            self.prepend(value)
            return
        new_node = Node(value)
        current = self.head
        for _ in range(position - 1):
            if current is None:
                raise IndexError("Position out of bounds")
            current = current.next
        if current is None:
            raise IndexError("Position out of bounds")
        
        new_node.next = current.next
        new_node.prev = current
        if current.next:
            current.next.prev = new_node
        current.next = new_node

    def deleteFirst(self):
        if not self.head:
            return
        self.head = self.head.next
        if self.head:
            self.head.prev = None

    def deleteLast(self):
        if not self.head:
            return
        if not self.head.next:
            self.head = None
            return
        current = self.head
        while current.next:
            current = current.next
        current.prev.next = None

    def delete(self, position):
        if not self.head:
            return
        if position == 0:
            self.deleteFirst()
            return
        current = self.head
        for _ in range(position):
            if current is None:
                return
            current = current.next
        if current is None:
            return
        
        if current.prev:
            current.prev.next = current.next
        if current.next:
            current.next.prev = current.prev

    def search(self, value):
        current = self.head
        position = 0
        while current:
            if current.value == value:
                return position
            current = current.next
            position += 1
        return -1

    def print(self):
        current = self.head
        elements = []
        while current:
            elements.append(str(current.value))
            current = current.next
        print(" <-> ".join(elements) if elements else "Empty List")

if __name__ == "__main__":
    dll = DoublyLinkedList()
    dll.append(10)
    dll.append(20)
    dll.prepend(5)
    dll.insert(15, 2)
    dll.print()  # Output: 5 <-> 10 <-> 15 <-> 20
    dll.delete(2)
    dll.print()  # Output: 5 <-> 10 <-> 20
